src/sourcetransform.rs

```diff
@@ -218,54 +223,116 @@ impl From<Message> for proto::source_transform_response::Result {
 {
+    type SourceTransformFnStream = ReceiverStream<Result<SourceTransformResponse, Status>>;
+
     async fn source_transform_fn(
         &self,
-        request: Request<proto::SourceTransformRequest>,
-    ) -> Result<Response<proto::SourceTransformResponse>, Status> {
-        let request = request.into_inner();
+        request: Request<Streaming<proto::SourceTransformRequest>>,
+    ) -> Result<Response<Self::SourceTransformFnStream>, Status> {
+        let mut stream = request.into_inner();
         let handler = Arc::clone(&self.handler);
-        let handle = tokio::spawn(async move { handler.transform(request.into()).await });
-        let shutdown_tx = self.shutdown_tx.clone();
-        let cancellation_token = self.cancellation_token.clone();
-
-        // Wait for the handler to finish processing the request. If the server is shutting down(token will be cancelled),
-        // then return an error.
-        tokio::select! {
-            result = handle => {
-                match result {
-                    Ok(messages) => Ok(Response::new(proto::SourceTransformResponse {
-                        results: messages.into_iter().map(|msg| msg.into()).collect(),
-                    })),
-                    Err(e) => {
-                        tracing::error!("Error in source transform handler: {:?}", e);
-                        // Send a shutdown signal to the server to do a graceful shutdown because there was
-                        // a panic in the handler.
-                        shutdown_tx.send(()).await.expect("Sending shutdown signal to gRPC server");
-                        Err(Status::internal(SourceTransformerError(UserDefinedError(e.to_string())).to_string()))
+
+        let (tx, rx) =
+            mpsc::channel::<Result<SourceTransformResponse, Status>>(DEFAULT_CHANNEL_SIZE);
+
+        // do the handshake first to let the client know that we are ready to receive read requests.
+        let handshake_request = stream
+            .message()
+            .await
+            .map_err(|e| Status::internal(format!("handshake failed {}", e)))?
+            .ok_or_else(|| Status::internal("stream closed before handshake"))?;
+
+        if let Some(handshake) = handshake_request.handshake {
+            tx.send(Ok(SourceTransformResponse {
+                results: vec![],
+                id: "".to_string(),
+                handshake: Some(handshake),
+            }))
+            .await
+            .map_err(|e| Status::internal(format!("failed to send handshake response {}", e)))?;
+        } else {
+            return Err(Status::invalid_argument("Handshake not present"));
+        }
+
+        let handle: JoinHandle<Result<(), Error>> = tokio::spawn({
+            let shutdown_tx = self.shutdown_tx.clone();
+            let cancellation_token = self.cancellation_token.clone();
+            let tx = tx.clone();
+            async move {
+                loop {
+                    tokio::select! {
```

Comment:
also can we divide the code into smaller functions for readability?

Reply:
Can we further breakdown it into smaller functions to improve readability?